File: Behavioral/TemplateMethod/php/template.php
<?php

/**
 * Class Builder
 */
abstract class Builder
{

    // Template method
    final public function build()
    {
        $this->test();
        $this->lint();
        $this->assemble();
        $this->deploy();
    }

    /**
     * @return mixed
     */
    abstract public function test();

    /**
     * @return mixed
     */
    abstract public function lint();

    /**
     * @return mixed
     */
    abstract public function assemble();

    /**
     * @return mixed
     */
    abstract public function deploy();
}


/**
 * Class AndroidBuilder
 */
class AndroidBuilder extends Builder
{
    public function test()
    {
        echo 'Running android tests';
    }
    public function lint()
    {
        echo 'Linting the android code';
    }
    public function assemble()
    {
        echo 'Assembling the android build';
    }

    public function deploy()
    {
        echo 'Deploying android build to server';
    }
}

/**
 * Class IosBuilder
 */
class IosBuilder extends Builder
{
    public function test()
    {
        echo 'Running ios tests';
    }

    public function lint()
    {
        echo 'Linting the ios code';
    }

    public function assemble()
    {
        echo 'Assembling the ios build';
    }

    public function deploy()
    {
        echo 'Deploying ios build to server';
    }
}

$androidBuilder = new AndroidBuilder();
$androidBuilder->build();

// Output:
// Running android tests
// Linting the android code
// Assembling the android build
// Deploying android build to server

$iosBuilder = new IosBuilder();
$iosBuilder->build();

// Output:
// Running ios tests
// Linting the ios code
// Assembling the ios build
// Deploying ios build to server

File: Behavioral/TemplateMethod/swift/template.swift
// Defining Builder protocol

protocol Builder {
    func test()
    func lint()
    func assemble()
    func deploy()
}

extension Builder {
    func build() {
        self.test()
        self.lint()
        self.assemble()
        self.deploy()
    }
}

// Defining Classes that implment Builder protocol

class AndroidBuilder: Builder {
    func test() {
        print("Running android tests")
    }

    func lint() {
        print("Linting the android code")
    }

    func assemble() {
        print("Assembling the android build")
    }

    func deploy() {
        print("Deploying android build to Google Play")
    }
}

class IosBuilder: Builder {
    func test() {
        print("Running iOS tests")
    }

    func lint() {
        print("Linting the iOS code")
    }

    func assemble() {
        print("Assembling the iOS build")
    }

    func deploy() {
        print("Deploying iOS build to AppStore")
    }
}

// Initialize Builders

AndroidBuilder().build()

// Output
// ------
// Running android tests
// Linting the android code
// Assembling the android build
// Deploying android build to Google Play

IosBuilder().build()

// Output
// ------
// Running iOS tests
// Linting the iOS code
// Assembling the iOS build
// eploying iOS build to AppStore
